catwalk/models/gpt.py

```diff
@@ -1,34 +1,161 @@
+from typing import Any, Dict, Iterator, List, Sequence, Tuple
 
 import more_itertools
 import torch
 from tango.common import Tqdm
+from tango.common.sequences import MappedSequence
 from tango.integrations.torch.util import resolve_device
-from torch import log_softmax
-from torch.nn.utils.rnn import pad_sequence
-from transformers import AutoModelForCausalLM, AutoTokenizer
 
 from catwalk import cached_transformers
-from catwalk.task import Task
 from catwalk.model import Model
+from catwalk.task import InstanceFormat, Task
+from catwalk.tasks.huggingface import HFQAInstance
+
+from torch import log_softmax
+from torch.nn.utils.rnn import pad_sequence
+from transformers import AutoModelForCausalLM, AutoTokenizer
 
 
 @Model.register("catwalk::gpt")
 class GPTModel(Model):
     def __init__(self, pretrained_model_name_or_path: str):
         self.pretrained_model_name_or_path = pretrained_model_name_or_path
+        
+    def _convert_instances(self, instances: Sequence[Dict[str, Any]], instance_format, task) -> MappedSequence:
+        return MappedSequence(lambda instance: task.convert_instance(instance, instance_format), instances)
 
     def predict(  # type: ignore
+        self,
+        task: Task,
+        instances: Sequence[Dict[str, Any]],
+        *,
+        batch_size: int = 32,
+    ) -> Iterator[Dict[str, Any]]:
+        if task.has_instance_conversion(InstanceFormat.HF_QA):
+            return self._predict_qa(task, instances, batch_size=batch_size)
+
+        return self._predict_preplixity(task, instances, batch_size=batch_size)
+
+    def _predict_qa(
+            self,
+            task: Task,
+            instances: Sequence[Dict[str, Any]],
+            *,
+            batch_size: int = 32
+    ) -> Iterator[Dict[str, Any]]:
+        device = resolve_device()
+        model = cached_transformers.get(
+            AutoModelForCausalLM, self.pretrained_model_name_or_path, False).eval().to(device)
+        tokenizer = cached_transformers.get_tokenizer(
+            AutoTokenizer, self.pretrained_model_name_or_path)
+        tokenizer.pad_token = tokenizer.eos_token
+        tokenizer.padding_side = 'left'
+        
+        converted_instances = self._convert_instances(instances, InstanceFormat.HF_QA, task)
+
+        def format_instance(instance: HFQAInstance) -> Tuple[str, str]:
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="max_length",
+                                  truncation="only_first",
+                                  max_length=384,  # TODO: Make this configurable?
```

Comment:
Yah, or perhaps let's just set it from the model config?
```suggestion
                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],  # TODO: Make this configurable?
```